### PR TITLE
Rewrite note field UI

### LINKED ISSUE #70: Rewrite note field
# 📖 Overview

The note field as-is will probably work fine but I don't like how it was organized and I think it could be more readable. Major goals for the rewrite are to split things up into smaller classes that make it easier to test and overall just improve the readability.

### PR DESCRIPTION
**Pull request by Kangaroux**
Closes #70 

The purpose of this rewrite is to make sure we are doing things the right way. The old code was bulky, awkward, used some visual tricks, etc. I tried to use native Gtk as much as possible.

There is a minor visual bug when the note field first loads but I'll take care of that in a separate issue

### CODE REVIEW

File: OpenChart/src/UI/NoteField/BeatLines.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        /// </summary>
+        public BeatLineSettings BeatLineSettings { get; private set; }
+
+        /// <summary>
+        /// The settings for the note field.
+        /// </summary>
+        public NoteFieldSettings NoteFieldSettings { get; private set; }
+
+        DrawingArea drawingArea;
+        public Widget GetWidget() => drawingArea;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/BeatLines.cs:23]: xmldocs for this getter?

File: OpenChart/src/UI/NoteField/BeatLines.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            cr.SetSourceColor(color.AsCairoColor());
+            cr.LineWidth = thickness;
+
+            // Fix the line not being the right thickness if the thickness is an odd number.
+            var applyOffset = (thickness % 2 == 1);
+
+            foreach (var y in positions)
+            {
+                var offsetY = y + (applyOffset ? 0.5 : 0);
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/BeatLines.cs:56]: maybe move the applyOffset logic to the variable? the resulting code could look something like this:

```cs
var offset = ((thickness % 2 == 1) ? 0.5 : 0)
foreach (var y in positions)
{
    var offsetY = y + offset;
    [...]
}
```

File: OpenChart/src/UI/NoteField/Key.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        /// </summary>
+        public NoteFieldSettings NoteFieldSettings { get; private set; }
+
+        /// <summary>
+        /// The key index in the chart that this widget represents.
+        /// </summary>
+        public KeyIndex KeyIndex { get; private set; }
+
+        SortedContainer<Beat> container;
+        public Gtk.Widget GetWidget() => container;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Key.cs:25]: same here

File: OpenChart/src/UI/NoteField/NoteField.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        /// </summary>
+        public KeyContainer Keys { get; private set; }
+
+        /// <summary>
+        /// The settings for the note field.
+        /// </summary>
+        public NoteFieldSettings NoteFieldSettings { get; private set; }
+
+        SortedContainer<Components> container;
+        public Gtk.Widget GetWidget() => container;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/NoteField.cs:36]: same here

File: OpenChart/src/UI/NoteField/NoteFieldView.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    /// </summary>
+    public class NoteFieldView : IWidget
+    {
+        /// <summary>
+        /// The note field widget.
+        /// </summary>
+        public NoteField NoteField { get; private set; }
+
+        public Scrollable Container { get; private set; }
+        public Widget GetWidget() => Container.GetWidget();
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/NoteFieldView.cs:16]: same here. also I notice in this case you chose to import Gtk, remove for consistency?

File: OpenChart/src/UI/NoteField/Objects/HoldNote.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    /// </summary>
+    public class HoldNote : INoteFieldObject
+    {
+        public int GetHeight() => Head.ImageAsset.Pixbuf.Height;
+
+        Charting.Objects.HoldNote chartObject;
+        public Charting.Objects.BaseObject GetChartObject() => chartObject;
+
+        SortedContainer<int> container;
+        public Widget GetWidget() => container;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/HoldNote.cs:18]: same

File: OpenChart/src/UI/NoteField/Objects/HoldNote.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        /// The body of the hold note.
+        /// </summary>
+        public HoldNoteBody Body { get; private set; }
+
+        /// <summary>
+        /// The head of the hold note.
+        /// </summary>
+        public HoldNoteHead Head { get; private set; }
+
+        public NoteFieldSettings NoteFieldSettings { get; private set; }
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/HoldNote.cs:30]: add xmldocs for this property

File: OpenChart/src/UI/NoteField/Objects/HoldNoteBody.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    public class HoldNoteBody : IWidget
+    {
+        /// <summary>
+        /// The image pattern used to display the body.
+        /// </summary>
+        public ImagePattern BodyImage { get; private set; }
+
+        DrawingArea drawingArea;
+        public Widget GetWidget() => drawingArea;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/HoldNoteBody.cs:17]: same

File: OpenChart/src/UI/NoteField/Objects/HoldNoteHead.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+{
+    /// <summary>
+    /// Widget for the head of a hold note.
+    /// </summary>
+    public class HoldNoteHead : IWidget
+    {
+        public ImageAsset ImageAsset { get; private set; }
+
+        Image widget;
+        public Gtk.Widget GetWidget() => widget;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/HoldNoteHead.cs:14]: same

File: OpenChart/src/UI/NoteField/Objects/NoteFieldObjectFactory.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+        /// <summary>
+        /// Creates a new NoteFieldObjectFactory instance.
+        /// </summary>
+        public NoteFieldObjectFactory(NoteFieldSettings noteFieldSettings)
+        {
+            NoteFieldSettings = noteFieldSettings;
+        }
+
+        public INoteFieldObject Create(Charting.Objects.BaseObject chartObject)
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/NoteFieldObjectFactory.cs:21]: add XMLdocs for this method

File: OpenChart/src/UI/NoteField/Objects/TapNote.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /// <summary>
+    /// Represents a tap note that is displayed on the note field.
+    /// </summary>
+    public class TapNote : INoteFieldObject
+    {
+        public int GetHeight() => ImageAsset.Pixbuf.Height;
+
+        Charting.Objects.TapNote chartObject;
+        public Charting.Objects.BaseObject GetChartObject() => chartObject;
```
Comment (monotron) [inline OpenChart/src/UI/NoteField/Objects/TapNote.cs:14]: same

File: OpenChart/src/UI/Widgets/Scrollable.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace OpenChart.UI.NoteField
+{
+    /// <summary>
+    /// A container which wraps a widget and provides horizontal and vertical scrolling.
+    /// </summary>
+    public class Scrollable : IWidget
+    {
+        Layout layout;
+        public Widget GetWidget() => layout;
```
Comment (monotron) [inline OpenChart/src/UI/Widgets/Scrollable.cs:12]: same

File: OpenChart/src/UI/NoteField/BeatLines.cs (hunk 2 above)
Comment (Kangaroux) [reply OpenChart/src/UI/NoteField/BeatLines.cs:56]: That's a 2AM moment for sure

Comment (monotron) [reply OpenChart/src/UI/NoteField/BeatLines.cs:56]: mood

**Pull request merged**